### PR TITLE
New map assertions

### LINKED ISSUE #944: Using Conditions to assert maps entries/keys/values
I wonder if it makes sense to make a next step from #205 and #690 and add more methods to `MapAssert`. 

Like:
```java
hasEntry(Condition<? super E> entryCondition)
hasValue(Condition<? super V> valueCondition)
hasKey(Condition<? super K> keyCondition)
```
and their negative counterparts.

### PR DESCRIPTION
**Pull request by fiery-phoenix**
#### Check List:
* Fixes #944 
* Unit tests : YES
* Javadoc with a code example (API only) : YES

Added:

    hasEntrySatisfying(Condition<? super Map.Entry<K, V>> entryCondition)
    hasEntrySatisfying(Condition<? super K> keyCondition, Condition<? super V> valueCondition)
    hasKeySatisfying(Condition<? super K> keyCondition)
    hasValueSatisfying(Condition<? super V> valueCondition)

### CODE REVIEW

**Comment by PascalSchumacher** [comment]
Great pull request! :+1:

**Comment by fiery-phoenix** [comment]
:)

File: src/main/java/org/assertj/core/api/AbstractMapAssert.java [hunk 1]
```diff
@@ -396,6 +396,180 @@ public SELF hasEntrySatisfying(K key, Condition<? super V> valueCondition) {
...
+   *     public boolean matches(TolkienCharacter value) {
+   *       return value.getRace() == ORC;
+   *     }
+   *   };
+   *
+   * // assertion will fail
+   * assertThat(ringBearers).hasKeySatisfying(isOrc);
+   *
+   * // assertion will pass
+   * assertThat(ringBearers).hasKeySatisfying(isMan);
```
Comment (joel-costigliola) [inline src/main/java/org/assertj/core/api/AbstractMapAssert.java]: that actually fails as galadriel and elrond are elves, frodo is a hobbit and gandalf is not a man but a maia :smile_cat:

**Comment by joel-costigliola** [comment]
Integrated, this PR was really good, I have made small changes improving a bit error messages and fixing the javadoc :D (as per my PR comment)

File: src/main/java/org/assertj/core/api/AbstractMapAssert.java (hunk 1 above)
Comment (fiery-phoenix) [inline src/main/java/org/assertj/core/api/AbstractMapAssert.java]: Forgot to include Isildur here. Thank you!

**Pull request closed without merging**